neotest-kotlin/app/src/main/kotlin/io/github/codymikol/neotestkotlin/framework/TestFrameworkRunner.kt

```diff
@@ -0,0 +1,52 @@
+package io.github.codymikol.neotestkotlin.framework
+
+import io.github.codymikol.neotestkotlin.framework.kotest.KotestTestRunner
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.fold
+import kotlinx.coroutines.flow.map
+import kotlin.reflect.KClass
+
+/**
+ * Base interface for all test framework runners.
+ */
+interface TestFrameworkRunner {
+    /**
+     * Runs the provided test [classes] using the [TestFrameworkRunner].
+     */
+    suspend fun run(classes: Collection<KClass<*>>): TestRunResult
+
+    /**
+     * Whether this class is runnable by this [TestFrameworkRunner].
+     */
+    fun isRunnable(kclass: KClass<*>): Boolean
+
+    companion object {
+        /**
+         * Executes all tests using all supported [TestFrameworkRunner]s
+         * generating a [RunReport] that contains all classes and their corresponding test
+         * statuses.
+         */
+        suspend fun runAll(classes: List<KClass<*>>): RunReport =
+            flowOf<TestFrameworkRunner>(
+                KotestTestRunner,
+            ).map { runner ->
+                val runnableClasses = classes.filter { runner.isRunnable(it) }
+
+                val result = runner.run(runnableClasses)
+                when (result) {
+                    is TestRunResult.Success -> result.report
+                    is TestRunResult.Failure -> TODO()
```

Comment:
No idea what we should do when this fails, I imagine we want to spit out some JSON that explains what went wrong here?

Reply:
I think that makes sense, on the lua side we can probably handle that by printing that into the overall output and failing all test cases with a generic error that basically says to look at the output for more information.

It doesn't look like you'd normally get this output and that it would either be a mistake on our end, or an internal kotest error.